<a href="https://colab.research.google.com/github/devesssi/EDUNNET/blob/main/logistic(work).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Load the data

### Subtask:
Load the dataset from "/content/Crop_and_fertilizer_with_usage.csv" into a pandas DataFrame.


**Reasoning**:
Import the pandas library, read the CSV file into a DataFrame, and display the head and info of the DataFrame to understand its structure and content.



In [15]:
import pandas as pd

df = pd.read_csv('/content/Crop_and_fertilizer_with_usage.csv')

display(df.head())
display(df.info())

District_Name Soil_color  Nitrogen  Phosphorus  Potassium   pH  Rainfall  \
0      Kolhapur      Black        75          50        100  6.5      1000   
1      Kolhapur      Black        80          50        100  6.5      1000   
2      Kolhapur      Black        85          50        100  6.5      1000   
3      Kolhapur      Black        90          50        100  6.5      1000   
4      Kolhapur      Black        95          50        100  6.5      1000   

   Temperature       Crop Fertilizer                          Link  \
0           20  Sugarcane       Urea  https://youtu.be/2t5Am0xLTOo   
1           20  Sugarcane       Urea  https://youtu.be/2t5Am0xLTOo   
2           20  Sugarcane       Urea  https://youtu.be/2t5Am0xLTOo   
3           20  Sugarcane       Urea  https://youtu.be/2t5Am0xLTOo   
4           20  Sugarcane       Urea  https://youtu.be/2t5Am0xLTOo   

   Fertilizer_Usage  
0             68.86  
1             88.52  
2             84.69  
3             83.16  
4             71.15

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4513 entries, 0 to 4512
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   District_Name     4513 non-null   object 
 1   Soil_color        4513 non-null   object 
 2   Nitrogen          4513 non-null   int64  
 3   Phosphorus        4513 non-null   int64  
 4   Potassium         4513 non-null   int64  
 5   pH                4513 non-null   float64
 6   Rainfall          4513 non-null   int64  
 7   Temperature       4513 non-null   int64  
 8   Crop              4513 non-null   object 
 9   Fertilizer        4513 non-null   object 
 10  Link              4513 non-null   object 
 11  Fertilizer_Usage  4513 non-null   float64
dtypes: float64(2), int64(5), object(5)
memory usage: 423.2+ KB


None

## Prepare the data

### Subtask:
Separate the features (input) and the target (output) variables. Remove the specified columns from the features.


**Reasoning**:
Separate the features (input) and the target (output) variables as instructed, and display the first few rows of each to verify.



In [16]:
y = df['Crop']
X = df.drop(['Crop', 'Fertilizer', 'Link'], axis=1)

display(X.head())
display(y.head())

District_Name Soil_color  Nitrogen  Phosphorus  Potassium   pH  Rainfall  \
0      Kolhapur      Black        75          50        100  6.5      1000   
1      Kolhapur      Black        80          50        100  6.5      1000   
2      Kolhapur      Black        85          50        100  6.5      1000   
3      Kolhapur      Black        90          50        100  6.5      1000   
4      Kolhapur      Black        95          50        100  6.5      1000   

   Temperature  Fertilizer_Usage  
0           20             68.86  
1           20             88.52  
2           20             84.69  
3           20             83.16  
4           20             71.15

0    Sugarcane
1    Sugarcane
2    Sugarcane
3    Sugarcane
4    Sugarcane
Name: Crop, dtype: object

## Split the data

### Subtask:
Split the dataset into training and testing sets to evaluate the model's performance.


**Reasoning**:
Split the features and target into training and testing sets.



In [17]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (3610, 9)
Shape of X_test: (903, 9)
Shape of y_train: (3610,)
Shape of y_test: (903,)


## Build and train the logistic regression model

### Subtask:
Create a Logistic Regression model and train it using the training data.


**Reasoning**:
Create and train a Logistic Regression model using the training data.



In [18]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Identify categorical and numerical features
categorical_features = ['District_Name', 'Soil_color']
numerical_features = ['Nitrogen', 'Phosphorus', 'Potassium', 'pH', 'Rainfall', 'Temperature', 'Fertilizer_Usage']

# Create a column transformer to apply one-hot encoding to categorical features
preprocessor = ColumnTransformer(
    transformers=[
        ('num', 'passthrough', numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

# Create a pipeline with the preprocessor and the logistic regression model
logistic_model = Pipeline(steps=[('preprocessor', preprocessor),
                               ('classifier', LogisticRegression(max_iter=1000))])

# Train the model
logistic_model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', 'passthrough',
                                                  ['Nitrogen', 'Phosphorus',
                                                   'Potassium', 'pH',
                                                   'Rainfall', 'Temperature',
                                                   'Fertilizer_Usage']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['District_Name',
                                                   'Soil_color'])])),
                ('classifier', LogisticRegression(max_iter=1000))])

**Reasoning**:
The error indicates that there are non-numeric values in the input features (X_train) which need to be handled before training the logistic regression model. I need to encode the categorical features.



In [19]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

categorical_features = ['District_Name', 'Soil_color']
numerical_features = ['Nitrogen', 'Phosphorus', 'Potassium', 'pH', 'Rainfall', 'Temperature', 'Fertilizer_Usage']

preprocessor = ColumnTransformer(
    transformers=[
        ('num', 'passthrough', numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

logistic_model = Pipeline(steps=[('preprocessor', preprocessor),
                               ('classifier', LogisticRegression(max_iter=1000))])

logistic_model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', 'passthrough',
                                                  ['Nitrogen', 'Phosphorus',
                                                   'Potassium', 'pH',
                                                   'Rainfall', 'Temperature',
                                                   'Fertilizer_Usage']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['District_Name',
                                                   'Soil_color'])])),
                ('classifier', LogisticRegression(max_iter=1000))])

## Evaluate the logistic regression model

### Subtask:
Evaluate the performance of the Logistic Regression model on the testing data using appropriate metrics like accuracy and a classification report.


**Reasoning**:
Evaluate the performance of the Logistic Regression model on the testing data using appropriate metrics like accuracy and a classification report.



In [20]:
from sklearn.metrics import classification_report, accuracy_score
from sklearn.linear_model import LogisticRegression

y_pred = logistic_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("\nClassification Report:\n", report)

Accuracy: 0.8283499446290143

Classification Report:
               precision    recall  f1-score   support

      Cotton       0.90      0.86      0.88       146
      Ginger       0.96      0.85      0.90        26
        Gram       0.50      0.21      0.30        14
      Grapes       1.00      1.00      1.00        19
   Groundnut       0.81      0.71      0.75        41
       Jowar       0.56      0.73      0.64        66
       Maize       0.94      0.90      0.92        72
      Masoor       0.00      0.00      0.00         2
       Moong       0.62      0.62      0.62        13
        Rice       0.84      0.73      0.78        64
     Soybean       0.67      0.50      0.57         8
   Sugarcane       0.93      0.96      0.94       205
         Tur       0.50      0.52      0.51        21
    Turmeric       0.50      0.50      0.50        12
        Urad       0.42      0.28      0.33        18
       Wheat       0.82      0.91      0.86       176

    accuracy              

## Build and train the decision tree model (optional)

### Subtask:
Build and train the decision tree model using the training data.


**Reasoning**:
Build and train the decision tree model using the training data by creating a pipeline that includes the preprocessor used for the logistic regression model and a DecisionTreeClassifier, then fitting this pipeline to the training data.



In [21]:
from sklearn.tree import DecisionTreeClassifier

decision_tree_model = Pipeline(steps=[('preprocessor', preprocessor),
                                    ('classifier', DecisionTreeClassifier(random_state=42))])

decision_tree_model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', 'passthrough',
                                                  ['Nitrogen', 'Phosphorus',
                                                   'Potassium', 'pH',
                                                   'Rainfall', 'Temperature',
                                                   'Fertilizer_Usage']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['District_Name',
                                                   'Soil_color'])])),
                ('classifier', DecisionTreeClassifier(random_state=42))])

## Evaluate the decision tree model (optional)

### Subtask:
Evaluate the performance of the Decision Tree model on the testing data and compare it with the Logistic Regression model.


**Reasoning**:
Make predictions with the Decision Tree model, calculate accuracy, generate a classification report, and print the results.



In [22]:
from sklearn.metrics import classification_report, accuracy_score

y_pred_dt = decision_tree_model.predict(X_test)

accuracy_dt = accuracy_score(y_test, y_pred_dt)
report_dt = classification_report(y_test, y_pred_dt)

print("Decision Tree Model Performance:")
print("Accuracy:", accuracy_dt)
print("\nClassification Report:\n", report_dt)

print("\nLogistic Regression Model Accuracy:", accuracy)
print("\nLogistic Regression Classification Report:\n", report)

Decision Tree Model Performance:
Accuracy: 0.9944629014396457

Classification Report:
               precision    recall  f1-score   support

      Cotton       1.00      1.00      1.00       146
      Ginger       1.00      0.96      0.98        26
        Gram       1.00      0.86      0.92        14
      Grapes       1.00      1.00      1.00        19
   Groundnut       0.93      1.00      0.96        41
       Jowar       1.00      0.98      0.99        66
       Maize       1.00      1.00      1.00        72
      Masoor       1.00      1.00      1.00         2
       Moong       1.00      1.00      1.00        13
        Rice       1.00      1.00      1.00        64
     Soybean       1.00      0.88      0.93         8
   Sugarcane       1.00      1.00      1.00       205
         Tur       1.00      1.00      1.00        21
    Turmeric       1.00      1.00      1.00        12
        Urad       0.95      1.00      0.97        18
       Wheat       1.00      1.00      1.00     

## Summary:

### Data Analysis Key Findings

*   The dataset contains 4513 entries and 12 columns with no missing values.
*   The Logistic Regression model achieved an accuracy of approximately 0.83 on the test set.
*   The Logistic Regression model showed high precision and recall for some classes (e.g., 'Grapes', 'Sugarcane') but struggled with others (e.g., 'Gram', 'Masoor', 'Urad'), resulting in lower performance metrics.
*   The Decision Tree model achieved a significantly higher accuracy of approximately 0.994 on the test set compared to the Logistic Regression model.
*   The Decision Tree model performed exceptionally well across almost all crop types, with high precision, recall, and F1-scores for most classes.

### Insights or Next Steps

*   The Decision Tree model is significantly better at predicting the crop to grow based on the provided features than the Logistic Regression model.
*   Further investigation into the classes where the Logistic Regression model performed poorly could involve exploring class imbalance or feature engineering to improve its performance, although the Decision Tree model already provides excellent results.


**Explanation of Code Cell `7dbfbdfb`:**

This cell imports the `pandas` library and reads the CSV file `/content/Crop_and_fertilizer_with_usage.csv` into a DataFrame named `df`. It then displays the first 5 rows of the DataFrame using `df.head()` and information about the DataFrame's columns and data types using `df.info()`. This helps in understanding the structure and content of the dataset.

**Explanation of Code Cell `c28a7793`:**

This cell separates the features (`X`) and the target variable (`y`). The target variable `y` is assigned the 'Crop' column from the DataFrame. The features `X` are created by dropping the 'Crop', 'Fertilizer', and 'Link' columns from the original DataFrame `df` using the `drop()` method with `axis=1` to indicate dropping columns. The first few rows of both `X` and `y` are then displayed to verify the separation.

**Explanation of Code Cell `fa4a1464`:**

This cell imports the `train_test_split` function from `sklearn.model_selection` and splits the features (`X`) and target variable (`y`) into training and testing sets. `test_size=0.2` specifies that 20% of the data will be used for testing, and `random_state=42` ensures reproducibility of the split. The shapes of the resulting training and testing sets are then printed to confirm the split.

**Explanation of Code Cell `nkuzYgK2XxP8`:**

This cell builds and trains the logistic regression model.
- It imports necessary classes: `OneHotEncoder` for handling categorical features, `ColumnTransformer` to apply different transformations to different columns, and `Pipeline` to chain preprocessing and modeling steps.
- It defines lists of categorical and numerical features.
- A `ColumnTransformer` is created:
    - `'num', 'passthrough', numerical_features` tells the transformer to leave the numerical features as they are.
    - `'cat', OneHotEncoder(handle_unknown='ignore'), categorical_features` tells the transformer to apply one-hot encoding to the categorical features. `handle_unknown='ignore'` is used to handle unseen categories during testing by ignoring them.
- A `Pipeline` is created with two steps: the `preprocessor` and a `LogisticRegression` model with `max_iter=1000` to ensure convergence.
- The `pipeline.fit(X_train, y_train)` line trains the entire pipeline (preprocessing and logistic regression model) using the training data.

**Explanation of Code Cell `d3f26370`:**

This cell evaluates the trained logistic regression model.
- It imports `classification_report` and `accuracy_score` from `sklearn.metrics`.
- `logistic_model.predict(X_test)` makes predictions on the testing data.
- `accuracy_score(y_test, y_pred)` calculates the accuracy of the model's predictions compared to the actual test labels.
- `classification_report(y_test, y_pred)` generates a detailed report including precision, recall, F1-score, and support for each class.
- The calculated accuracy and the classification report are then printed.

**Explanation of Code Cell `18cab710`:**

This cell builds and trains the decision tree model.
- It imports the `DecisionTreeClassifier` from `sklearn.tree`.
- A `Pipeline` is created, similar to the logistic regression model, but with a `DecisionTreeClassifier` as the classifier step. `random_state=42` is set for reproducibility.
- `decision_tree_model.fit(X_train, y_train)` trains the decision tree pipeline using the training data.

**Explanation of Code Cell `398af67d`:**

This cell evaluates the trained decision tree model and compares its performance to the logistic regression model.
- It imports `classification_report` and `accuracy_score` from `sklearn.metrics`.
- `decision_tree_model.predict(X_test)` makes predictions on the testing data using the decision tree model.
- `accuracy_score(y_test, y_pred_dt)` calculates the accuracy for the decision tree model.
- `classification_report(y_test, y_pred_dt)` generates a classification report for the decision tree model.
- The accuracy and classification report for the decision tree model are printed.
- Finally, the accuracy and classification report from the logistic regression model (calculated in cell `d3f26370`) are also printed again for easy comparison.